File: python/test_local_search.py
#!/usr/bin/env python3
"""test_local_search.py - Unit test cho 2-opt / Or-opt.
Chạy: python3 -m pytest test_local_search.py -v
"""
import numpy as np
import ga_core as ga
import local_search as ls


def _perm_ok(t, n):
    return sorted(np.asarray(t).tolist()) == list(range(n))


def test_two_opt_keeps_permutation():
    rng = np.random.default_rng(0)
    coords = rng.random((15, 2))
    D = ga.distance_matrix(coords)
    for _ in range(20):
        t = ga.random_tour(15, rng)
        out = ls.two_opt(t, D)
        assert _perm_ok(out, 15)


def test_two_opt_not_worse():
    # 2-opt không bao giờ làm tour dài hơn
    rng = np.random.default_rng(1)
    coords = rng.random((20, 2))
    D = ga.distance_matrix(coords)
    for _ in range(20):
        t = ga.random_tour(20, rng)
        before = ga.tour_length(t, D)
        after = ga.tour_length(ls.two_opt(t, D), D)
        assert after <= before + 1e-9


def test_two_opt_fixes_crossing():
    # 4 điểm vuông: thứ tự chéo 0-2-1-3 (có giao) phải được 2-opt sửa về chu vi = 4
    coords = np.array([[0, 0], [0, 1], [1, 1], [1, 0]], dtype=float)
    D = ga.distance_matrix(coords)
    crossed = np.array([0, 2, 1, 3])
    fixed = ls.two_opt(crossed, D)
    assert abs(ga.tour_length(fixed, D) - 4.0) < 1e-9


def test_or_opt_not_worse():
    rng = np.random.default_rng(2)
    coords = rng.random((18, 2))
    D = ga.distance_matrix(coords)
    for _ in range(15):
        t = ga.random_tour(18, rng)
        before = ga.tour_length(t, D)
        after = ga.tour_length(ls.or_opt(t, D, seg_len=2), D)
        assert after <= before + 1e-9


def test_polish_improves_random():
    rng = np.random.default_rng(3)
    coords = rng.random((25, 2))
    D = ga.distance_matrix(coords)
    t = ga.random_tour(25, rng)
    assert ga.tour_length(ls.polish(t, D), D) < ga.tour_length(t, D)


if __name__ == "__main__":
    fns = [v for k, v in sorted(globals().items()) if k.startswith("test_")]
    for fn in fns:
        fn()
        print(f"OK  {fn.__name__}")
    print(f"\nTat ca {len(fns)} test PASS.")

File: cpp/test_local_search.cpp
// test_local_search.cpp - Unit test for 2-opt / Or-opt (C++).
// Compile: g++ -O2 -o test_ls test_local_search.cpp && ./test_ls
#include "local_search.hpp"
#include <iostream>

static int passed = 0;
#define CHECK(cond, name) do { if (!(cond)) { \
    std::cerr << "FAIL " << name << "\n"; return 1; } \
    passed++; } while (0)

static bool is_perm(const Tour& t, int n) {
    std::vector<char> seen(n, 0);
    for (int c : t) { if (c < 0 || c >= n || seen[c]) return false; seen[c] = 1; }
    return (int)t.size() == n;
}

int main() {
    std::mt19937 rng(0);

    // 2-opt fixes the crossing back to perimeter 4
    std::vector<std::pair<double,double>> sq = {{0,0},{0,1},{1,1},{1,0}};
    auto Dsq = distance_matrix(sq);
    Tour crossed = {0, 2, 1, 3};
    two_opt(crossed, Dsq, 4);
    CHECK(std::abs(tour_length(crossed, Dsq, 4) - 4.0) < 1e-9, "two_opt_fixes_crossing");

    // 2-opt & or-opt do not make it worse, and preserve the permutation
    std::vector<std::pair<double,double>> pts;
    std::uniform_real_distribution<double> u(0, 1);
    for (int i = 0; i < 20; i++) pts.emplace_back(u(rng), u(rng));
    auto D = distance_matrix(pts);
    for (int rep = 0; rep < 30; rep++) {
        Tour t = random_tour(20, rng);
        double before = tour_length(t, D, 20);
        Tour t2 = t; two_opt(t2, D, 20);
        CHECK(is_perm(t2, 20), "two_opt_perm");
        CHECK(tour_length(t2, D, 20) <= before + 1e-9, "two_opt_not_worse");
        Tour t3 = t; or_opt(t3, D, 20);
        CHECK(is_perm(t3, 20), "or_opt_perm");
        CHECK(tour_length(t3, D, 20) <= before + 1e-9, "or_opt_not_worse");
    }

    // polish improves a random tour
    Tour t = random_tour(20, rng);
    double before = tour_length(t, D, 20);
    polish(t, D, 20);
    CHECK(tour_length(t, D, 20) < before, "polish_improves");

    std::cout << "All tests passed (" << passed << " checks).\n";
    return 0;
}
